### PR TITLE
#84/update firebase token

### PR DESCRIPTION
**Pull request by timosturm**
If push tokens exist in the app, the following is now possible:
+ the user can manually synchronize the firebase token of the app with the privacyIDEA server(s) as shown here:
Screenshot_20210106_140049_it netknights piauthenticator
Screenshot_20210106_140124_it netknights piauthenticator
Screenshot_20210106_140056_it netknights piauthenticator
Screenshot_20210106_140141_it netknights piauthenticator

+ If the firebase token is updated while the app is running, the app attempts to automatically synchronize the firebase token with the privacyIDEA server(s), if this fails, it attempts to perform this update at  every app start until success.

+ Minor fix: A Scrollbar is shown in the screen for manually adding tokens if the content can not be displayed at once.

+ Also added user agent string where appropriate

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -13,7 +13,7 @@ repository: https://github.com/privacyidea/pi-authenticator
 # In iOS, build-name is used as CFBundleShortVersionString while build-number used as CFBundleVersion.
 # Read more about iOS versioning at
 # https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html
-version: 3.0.8+0300008 # TODO Set the right version number
+version: 3.0.9+0300009 # TODO Set the right version number
```
Comment (nilsbehlen) [inline pubspec.yaml:16]: the current beta in the playstore has version 3.0.7 if im not mistaken. Therefore, i think we should set this to 3.0.8.

**Review by nilsbehlen — COMMENTED** [review]
see the comment about version, otherwise this looks good.

File: lib/widgets/update_firebase_token_dialog.dart [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import 'package:http/http.dart';
+import 'package:pi_authenticator_legacy/pi_authenticator_legacy.dart';
+import 'package:privacyidea_authenticator/model/tokens.dart';
+import 'package:privacyidea_authenticator/utils/crypto_utils.dart';
+import 'package:privacyidea_authenticator/utils/localization_utils.dart';
+import 'package:privacyidea_authenticator/utils/storage_utils.dart';
+import 'package:privacyidea_authenticator/utils/utils.dart';
+
+class UpdateFirebaseTokenDialog extends StatefulWidget {
+  UpdateFirebaseTokenDialog();
```
Comment (timosturm) [inline lib/widgets/update_firebase_token_dialog.dart:36]:
```suggestion
```

File: lib/widgets/update_firebase_token_dialog.dart [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        content: _content,
+        actions: <Widget>[_button],
+      ),
+    );
+  }
+
+  void _updateFbTokens() async {
+    log(
+      'Starting update of firebase token.',
+      name: 'update_firebase_token_dialog.dart',
```
Comment (timosturm) [inline lib/widgets/update_firebase_token_dialog.dart:71]:
```suggestion
      name: 'update_firebase_token_dialog.dart'
```

File: lib/widgets/update_firebase_token_dialog.dart [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        'serial': p.serial,
+        'timestamp': timestamp,
+        'signature': signature
+      });
+
+      if (response.statusCode == 200) {
+        log('Updating firebase token for push token: ${p.serial} succeeded!');
+      } else {
+        log('Updating firebase token for push token: ${p.serial} failed!',
+            name: 'main_screen.dart');
```
Comment (timosturm) [inline lib/widgets/update_firebase_token_dialog.dart:115]:
```suggestion
            name: 'update_firebase_token_dialog.dart');
```

File: lib/widgets/update_firebase_token_dialog.dart [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+      Response response =
+          await doPost(sslVerify: p.sslVerify, url: p.url, body: {
+        'new_fb_token': token,
+        'serial': p.serial,
+        'timestamp': timestamp,
+        'signature': signature
+      });
+
+      if (response.statusCode == 200) {
+        log('Updating firebase token for push token: ${p.serial} succeeded!');
```
Comment (timosturm) [inline lib/widgets/update_firebase_token_dialog.dart:112]:
```suggestion
        log('Updating firebase token for push token: ${p.serial} succeeded!',name: 'update_firebase_token_dialog.dart');
```

**Review by nilsbehlen — APPROVED** [review]

**Pull request merged**